lib.modules: Let module declare options directly in bare submodule

roberth opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Let module declare options directly in bare submodule, where a bare submodule is an option that has a type like `submoduleWith x`, as opposed to `attrsOf (submoduleWith x)`.

This makes migration unnecessary when introducing a freeform type in an existing option tree and it removes some unwieldy syntax. Refs https://github.com/NixOS/nixpkgs/pull/156503/files#r790768951

Mainly, it allows the introduction of a freeformType at any place in an existing options tree without breaking compatibility with existing option declarations inside the new freeformType.

Another use case is for modules that should be imported in multiple locations in the options tree, such as `systemd` and `systemd.user`.

Closes #146882

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

roberth committed:
lib.modules: Let module declare options directly in bare submodule

... where a bare submodule is an option that has a type like
`submoduleWith x`, as opposed to `attrsOf (submoduleWith x)`.

This makes migration unnecessary when introducing a freeform type
in an existing option tree.

Closes #146882

ncfavier commented on lib/modules.nix:
```diff
@@ -471,6 +471,18 @@ rec {
 
+      # Convert an option tree decl to a submodule option decl
+      optionTreeToOption = decl:
+        if builtins.trace decl isOption decl.options
```
Don't forget to remove the trace

ncfavier commented on lib/modules.nix:
```diff
@@ -490,7 +502,15 @@ rec {
-              throw "The option `${showOption loc}' in `${firstOption._file}' is a prefix of options in `${firstNonOption._file}'."
+              if firstOption.options.type?isSubmodule && firstOption.options.type.isSubmodule
+              then
+                let opt = fixupOptionType loc (mergeOptionDecls loc (map optionTreeToOption decls));
+                in {
+                  matchedOptions = evalOptionValue loc opt defns';
+                  unmatchedDefns = [];
+                }
```
This could probably be deduplicated with the code just above.

roberth replied:
Done.
It could be made even less repetitive, but that would require more conditionals or function calls that I'd prefer not to add because of evaluation performance.

ncfavier replied:
I think what I meant was that the first two cases can be combined into one as they return the same value (`map optionTreeToOption decls == decls` in the first case).

```nix
if optionDecls == [] then
  mergeModules' loc decls defns
else if (lib.head optionDecls).options.type.name != "submodule" then
  throw "The option ... is a prefix of ..."
else
  let opt = fixupOptionType loc (mergeOptionDecls loc (map optionTreeToOption decls));
  in {
    matchedOptions = evalOptionValue loc opt defns';
    unmatchedDefns = [];
  }
```

I don't think this hurts performance? It might slightly change the semantics if `decls == []`, but is that case even possible?

roberth replied:
> I guess it calls optionTreeToOption more.

Exactly. Calls are quite expensive.

> make `map optionTreeToOption` a fold

I don't expect going from two list function to a single currying list function to be an improvement. Having to then store both results as an intermediate value is very likely to make it more expensive.

> Does this kind of micro-optimisation really matter in practice?

Often it doesn't, but this code is called very frequently. OfBorg's total allocated bytes are down by 1/800th. I'd say that's not insignificant. I'd speculate based on the numbers that a `builtins.count` could achieve four times that, or 0.5%.

I don't think we need to improve performance any further in this PR, but It will be interesting to try something in a follow up, so we can see the effect of a single change.

ncfavier commented on lib/types.nix:
```diff
@@ -594,6 +594,11 @@ rec {
+        # Submodule-typed options get special treatment in order to facilitate
+        # certain migrations, such as the addition of freeformTypes onto
+        # existing option trees.
+        isSubmodule = true;
```
I think for further extensibility it would make sense to abstract this into a function, so that e.g. `listOf` can define its own logic for things like `foo."*".bar`. This would replace `optionTreeToOption` (or perhaps `map optionTreeToOption`) in `lib/modules`.

roberth replied:
I agree, but I can't decide this without also deciding what the semantics of, say, `attrsOf` should be when it comes to option merging. It's ok to take small steps. I've added an explanation as an inline comment.

I've removed `isSubmodule` is it wasn't strictly necessary. This way we won't have to remove it when we do generalize this feature.

roberth committed:
lib.types.submodule: Remove redundant isSubmodule attr

roberth committed:
lib.modules: Refactor option scanning slightly

This scans the options with fewer function calls, improving performance.

It also removes a let Env from the happy flow of the new logic.

roberth committed:
lib.modules: Explain why options can only be merged into submodules

Infinisil commented on lib/modules.nix:
```diff
@@ -475,26 +474,54 @@ rec {
-            in
-              throw "The option `${showOption loc}' in `${firstOption._file}' is a prefix of options in `${firstNonOption._file}'."
+          else if optionDecls != [] then
+              if (lib.head optionDecls).options.type.name == "submodule"
```
This leads to a different error depending on module order. E.g.
```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submoduleWith { modules = []; };
      };
    }
    {
      _file = "b.nix";
      options.foo = lib.mkOption {
        type = lib.types.int;
      };
    }
    {
      _file = "c.nix";
      options.foo.bar = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

gives

```
error: The option `foo' in `b.nix' is a prefix of options in `c.nix'.
(use '--show-trace' to show detailed location information)
```

While

```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "b.nix";
      options.foo = lib.mkOption {
        type = lib.types.int;
      };
    }
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submoduleWith { modules = []; };
      };
    }
    {
      _file = "c.nix";
      options.foo.bar = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

gives

```
error: The option `foo' in `b.nix' is already declared in `c.nix' and `a.nix'.
(use '--show-trace' to show detailed location information)
```

Not yet sure what the best error to throw is in such a case

roberth replied:
Yeah, I've inherited this flaw from the existing code.
Might have a look later, but not sure if this needs to be scoped in.

Infinisil replied:
Previously the order of modules only had an influence on the specific files mentioned in the error message, but the error message stayed the same. This PR would make the error message variety itself change.

roberth replied:
It now checks all and I've improved the error message.

Infinisil commented on lib/modules.nix:
```diff
@@ -475,26 +474,54 @@ rec {
+        then decl
+        else decl // {
+            options = mkOption {
+              type = types.submoduleWith {
+                modules = [ { options = decl.options; } ];
+              };
```
This leads to an error when a users option is declared with just `types.submodule`:

```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submodule {};
        default = {};
      };
    }
    {
      _file = "b.nix";
      options.foo.bar = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

gives

```
error: A submoduleWith option is declared multiple times with conflicting shorthandOnlyDefinesConfig values
(use '--show-trace' to show detailed location information)
```

Not sure if this can be fixed nicely or if it even makes sense.

roberth replied:
`submodule` is the legacy one, but yeah, not great. Seems like "don't care" needs to be representable and pick `true` as the default. `false` was probably a bad idea.

roberth replied:
Added `null` support, but only when merging.

Infinisil commented:
I wonder if it would be feasible for option trees (like `options.foo.bar.baz = mkOption { ... }`) to just be desugared to submodules (like `options.foo = mkOption { type = submodule { options.bar = mkOption { type = submodule { options.baz = mkOption { ... }; }; }; }; }`) and in the process cleaning up the code :thinking:

roberth committed:
lib.modules: Improve option-is-prefix error message

roberth committed:
lib.modules: Make option injection work when shorthandOnlyDefinesConfig

roberth commented:
> I wonder if it would be feasible for option trees to just be desugared to submodules and in the process cleaning up the code

It occurred to me too, but it wasn't clear to me whether it will be effective.
Calling back to "high-level" functions like `submodule` may not be great for performance, so it seems like it'd be a significant rethink. Not something I'll explore in this PR.

roberth committed:
lib.modules: Remove redundant fixupOptionType in option injection

Infinisil commented on lib/types.nix:
```diff
@@ -608,7 +608,11 @@ rec {
               else throw "A submoduleWith option is declared multiple times with the same specialArgs \"${toString (attrNames intersecting)}\"";
             shorthandOnlyDefinesConfig =
-              if lhs.shorthandOnlyDefinesConfig == rhs.shorthandOnlyDefinesConfig
+              if lhs.shorthandOnlyDefinesConfig == null
+              then rhs.shorthandOnlyDefinesConfig
+              else if rhs.shorthandOnlyDefinesConfig == null
+              then lhs.shorthandOnlyDefinesConfig
+              else if lhs.shorthandOnlyDefinesConfig == rhs.shorthandOnlyDefinesConfig
```
Since `null` is now an allowed value, we should also define its semantics and document it. The semantics should be that if this value is needed to make a decision and it's `null`, an error is thrown. This means that if you do any shorthand definitions, you should get an error, unless another option declaration specifies the value of `shorthandOnlyDefinesConfig`, e.g. whereas

```nix
let
  lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      options.foo = lib.mkOption {
        type = lib.types.submoduleWith {
          modules = [];
          shorthandOnlyDefinesConfig = null;
        };
      };

      config.foo.bar = 10;
    }
  ];
}
```

with this PR currently throws
```
error: value is null while a Boolean was expected

       at /home/infinisil/src/nixpkgs/lib/types.nix:538:23:

          537|           then setFunctionArgs (args: unify (value args)) (functionArgs value)
          538|           else unify (if shorthandOnlyDefinesConfig then { config = value; } else value);
             |                       ^
          539|
(use '--show-trace' to show detailed location information)
```

it should instead be something like

```
error: shorthandOnlyDefinesConfig is null
(use '--show-trace' to show detailed location information)
```

roberth replied:
> Since `null` is now an allowed value

I disagree. You'd have to read the implementation to find that null is acceptable in some situations.

To keep things simple and, I've decided to default it to `true`. I don't see much point in documenting this though.

roberth committed:
lib.modules: Default shorthandOnlyDefinesConfig to true when null

roberth committed:
lib.modules: Inline a private function

This should save about four calls per module.

Infinisil commented on lib/modules.nix:
```diff
@@ -474,26 +488,58 @@ rec {
+              #  d. magically combine (a) and (c).
+              # All of the above are merely syntax sugar though.
+              then
+                let opt = mergeOptionDecls loc (map optionTreeToOption decls);
```
After a lot of trial and error, trying to prove why `fixupOptionType` should be used here or not, I figured it out: It's needed for the sake of file locations in error messages. E.g. with

```nix
let lib = import ./lib;
in lib.evalModules {
  modules = [
    {
      _file = "a.nix";
      options.foo = lib.mkOption {
        type = lib.types.submodule {
          options.x = lib.mkOption {
            type = lib.types.int;
          };
        };
        default = {};
      };
    }
    {
      _file = "b.nix";
      options.foo.x = lib.mkOption {
        type = lib.types.str;
      };
    }
  ];
}
```

it currently produces this error:

```
$ nix-instantiate --eval -A config.foo.x
error: The option `foo.x' in `<unknown-file>' is already declared in `<unknown-file>'.
(use '--show-trace' to show detailed location information)
```

whereas with a `fixupOptionType` here, it's

```
error: The option `foo.x' in `a.nix' is already declared in `b.nix'.
(use '--show-trace' to show detailed location information)
```

would be good to have a test case for this

roberth replied:
Done.

Infinisil commented on lib/modules.nix:
```diff
@@ -474,26 +488,58 @@ rec {
+            options = mkOption {
+              type = types.submoduleWith {
+                modules = [ { options = decl.options; } ];
+                shorthandOnlyDefinesConfig = null;
```
This should have a comment explaining the reasoning behind this: How it's a bit hacky but needed as a default, that `null` is only an internally valid value, that it's only safe here because of the merging function patch and because it's always going to be merged with at least one other module.

In the future this might become obsolete with https://github.com/NixOS/nixpkgs/issues/162398

roberth replied:
Added

```
                # `null` is not intended for use by modules. It is an internal
                # value that means "whatever the user has declared elsewhere".
                # This might become obsolete with https://github.com/NixOS/nixpkgs/issues/162398
```

roberth committed:
Revert "lib.modules: Remove redundant fixupOptionType in option injection"

This reverts commit 6b077c47ff14cb9a4a8f5cb8986fa83ff626c732.

Thanks Infinisil for discovering this problem:

> After a lot of trial and error, trying to prove why fixupOptionType should
> be used here or not, I figured it out: It's needed for the sake of file
> locations in error messages.

roberth committed:
lib/modules.nix: Move comment to the actual legacy code

roberth committed:
lib/tests/modules: Add test case for duplicate option error file location

roberth committed:
lib/modules.nix: Add comment about internal shorthand null value

infinisil reviewed (approved):
Looking good!

bobvanderlinden commented:
Could you explain the example for systemd a bit?

Would this allow something like:

```
let
  units = {
    options = {
      units = mkOption { ... };
    };
    config = { ... };
  };
in
{
  options = {
    boot.systemd = mkOption {
      type = types.submoduleWith {
        modules = [ units services mounts ... ];
      };
    };
  };
}
```

(That would be amazing)

roberth commented:
@bobvanderlinden Yes, while this was already valid, it can now be mixed with "top-level" options like

```
{
  options.boot.systemd.units.foo = mkOption { };
}
```

This was not allowed to coexist with your example.
So the refactoring you want to do has become less risky and possible to do incrementally, without having to migrate all options to the submodule at once.

roberth commented:
tl;dr we can improve laziness by translating bare submodules to option trees rather than the other way around, or we can implement a special "import at prefix" that does only what we need.

Submodules are more than just option trees though, so a "prefixing" operation on modules seems like it'd need quite a lot of code.
A more restricted "import at prefix" feature that only deals with options and configs may be easier to implement, as it doesn't have to translate freeformTypes. Maybe I'm overestimating the complexity of translating all submodule functionality though.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/genodepkgs-extending-nixpkgs-nixos-to-genode/8779/7

roberth commented:
This PR reinterprets parts of the option tree based on information of explicitly declared options.
This could be generalized to not just the improvement of attrsets in the option tree, but also plain functions in the option tree.
Specifically, we could perhaps make the `optionType` itself responsible for interpreting the subexpressions of `options` for the purpose of static merging.

I call it static merging for two reasons:
 - It generalizes type merging. Note that types are static information. (generally! You can easily do a dependent type, but this is meant to be descriptive, not prescriptive use of the term)
 - It is static and can be used for declaring nested docs (e.g. docs in a submodule). This sets it apart from the `config` part, which is dynamic and not queried for docs.

Motivating example: https://github.com/hercules-ci/flake-parts/issues/142

Atry commented:
I wonder if we could allow for merging an option with its default values, no matter what the type is.
